lib/bolt/transport/ssh.rb

```diff
@@ -0,0 +1,144 @@
+require 'bolt/node/errors'
+require 'bolt/transport/ssh/connection'
+require 'json'
+require 'logging'
+require 'shellwords'
+
+module Bolt
+  module Transport
+    class SSH
+      attr_reader :logger
+
+      STDIN_METHODS       = %w[both stdin].freeze
+      ENVIRONMENT_METHODS = %w[both environment].freeze
+
+      def initialize(_config)
+        @logger = Logging.logger[self]
+
+        require 'net/ssh'
+        require 'net/scp'
+        begin
+          require 'net/ssh/krb'
+        rescue LoadError
+          logger.debug {
+            "Authentication method 'gssapi-with-mic' is not available"
+          }
+        end
+      end
+
+      def with_connection(target)
+        conn = Connection.new(target)
+        conn.connect
+        yield conn
+      ensure
+        conn.disconnect if conn
+      end
+
+      def running_as(connection, user)
+        target = connection.target
+        original_run_as = target.options[:run_as]
+        target.options[:run_as] = user
```

Comment:
I think it might be worth avoiding this by saving a "run_as" value on the connect or something. The less we mutate target objects the better.